### PR TITLE
sshexec: Make sudo an optional command based on the user provided in config file.

### PR DESCRIPTION
**Pull request by humblec**
set sudovar if its a non root user and execute gluster commands.
If its root user dont set the sudovar.

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

### CODE REVIEW

File: executors/sshexec/sshexec.go [hunk 1]
```diff
@@ -26,6 +26,8 @@ import (
 	"github.com/lpabon/godbc"
 )
 
+var Sudovar string
```
Comment (lpabon) [inline executors/sshexec/sshexec.go]: Sudovar should not be a global, it should be part of the SshExecutor object in line 31

**Comment by lpabon** [comment]
Also, we should do another more maintainable way ( like ansible ).  Remove the sudo from all commands as you have done, but then pass a boolean in [sshExec](https://github.com/heketi/heketi/blob/master/executors/sshexec/sshexec.go#L132) to [utils/ssh/ssh.go](https://github.com/heketi/utils/blob/master/ssh/ssh.go#L108).  Ssh.go would then insert a sudo command at [line 136](https://github.com/heketi/utils/blob/master/ssh/ssh.go#L136).

Note that you would need to make the change to ssh.go first because it is in a different repo:  https://github.com/heketi/utils .

Once that is checked in, we can change update this fix to use the new api and update the GLOCKFILE.

File: executors/sshexec/sshexec.go (hunk 1 above)
Comment (humblec) [inline executors/sshexec/sshexec.go]: I will send the PR with this change.

**Comment by humblec** [comment]
@lpabon I am bit confused with the last comment.  Do we need to do it with this patch or are you saying we will do it in more maintainable way later?

**Comment by lpabon** [comment]
@humblec this patch

**Comment by humblec** [comment]
@lpabon the travis CI is not working for the latest patch set. Not sure why its not triggering. Any thoughts on this ?

**Comment by humblec** [comment]
@lpabon with the latest patch set, the travis CI is triggered, however failed with below error.
"# github.com/heketi/heketi/executors/sshexec

executors/sshexec/sshexec.go:71: cannot use s (type *"github.com/heketi/utils/ssh".SshExec) as type Ssher in return argument:

	*"github.com/heketi/utils/ssh".SshExec does not implement Ssher (wrong type for ConnectAndExec method)

		have ConnectAndExec(string, []string, int) ([]string, error)

		want ConnectAndExec(string, []string, int, bool) ([]string, error)

# github.com/heketi/heketi/executors/sshexec

executors/sshexec/sshexec.go:71: cannot use s (type *"github.com/heketi/utils/ssh".SshExec) as type Ssher in return argument:

	*"github.com/heketi/utils/ssh".SshExec does not implement Ssher (wrong type for ConnectAndExec method)

		have ConnectAndExec(string, []string, int) ([]string, error)

		want ConnectAndExec(string, []string, int, bool) ([]string, error)

"
Looks like the "github.com/heketi/utils/ssh" latest merge ( with one more arg to ConnectAndExec) is not reflected when travis is in action. Any thought  on this Luis ?

**Comment by humblec** [comment]
@lpabon locally all the tests are passing. However Travis CI is failing. As noted above, the new ConnectAndExec() with an extra arg will work when there is latest utils/ssh.go . I am confused on which one to fix first ? . Can you please help here ?

File: executors/sshexec/brick.go [hunk 2]
```diff
@@ -91,14 +91,14 @@ func (s *SshExecutor) BrickCreate(host string,
...
+		fmt.Sprintf(" mount -o rw,inode64,noatime,nouuid %v %v", s.devnode(brick), mountpoint),
 
 		// Create a directory inside the formated volume for GlusterFS
-		fmt.Sprintf("sudo mkdir %v/brick", mountpoint),
+		fmt.Sprintf(" mkdir %v/brick", mountpoint),
 	}
 
 	// Execute commands
-	_, err := s.RemoteExecutor.RemoteCommandExecute(host, commands, 10)
+	_, err := s.RemoteExecutor.RemoteCommandExecute(host, commands, 10, s.usesudo)
```
Comment (lpabon) [inline executors/sshexec/brick.go]: The RemoteComandExecute should not take usesudo

File: executors/sshexec/sshexec.go [hunk 3]
```diff
@@ -27,11 +27,11 @@ import (
 )
 
 type RemoteCommandTransport interface {
-	RemoteCommandExecute(host string, commands []string, timeoutMinutes int) ([]string, error)
+	RemoteCommandExecute(host string, commands []string, timeoutMinutes int, useSudo bool) ([]string, error)
```
Comment (lpabon) [inline executors/sshexec/sshexec.go]: Don't add this one.

File: executors/sshexec/sshexec.go [hunk 4]
```diff
@@ -27,11 +27,11 @@ import (
...
 }
 
 type Ssher interface {
-	ConnectAndExec(host string, commands []string, timeoutMinutes int) ([]string, error)
+	ConnectAndExec(host string, commands []string, timeoutMinutes int, useSudo bool) ([]string, error)
```
Comment (lpabon) [inline executors/sshexec/sshexec.go]: :+1:

File: executors/sshexec/sshexec.go [hunk 5]
```diff
@@ -172,14 +179,14 @@ func (s *SshExecutor) FreeConnection(host string) {
...
-	timeoutMinutes int) ([]string, error) {
+	timeoutMinutes int, useSudo bool) ([]string, error) {
 
 	// Throttle
 	s.AccessConnection(host)
 	defer s.FreeConnection(host)
 
 	// Execute
-	return s.exec.ConnectAndExec(host+":"+s.port, commands, timeoutMinutes)
+	return s.exec.ConnectAndExec(host+":"+s.port, commands, timeoutMinutes, useSudo)
```
Comment (lpabon) [inline executors/sshexec/sshexec.go]: This should use s.usesudo instead.

**Comment by lpabon** [comment]
The GLOCKFILE needs to be updated in this patch to grab the changes from utils

**Comment by humblec** [comment]
@lpabon I have addressed all the comments, however the travis build has not triggered. Can you please check?

**Comment by humblec** [comment]
@lpabon I have a different PR (https://github.com/heketi/heketi/pull/353) as you mentioned and it triggered the travis CI and its passed. As soon as its merged, I will close this PR.

**Comment by humblec** [comment]
I am closing this PR as this has been addressed here https://github.com/heketi/heketi/pull/353

**Pull request closed without merging**